src/utils/src/backend-supports/project-supports.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * This file is part of NER's PM Dashboard and licensed under GNU AGPLv3.
+ * See the LICENSE file in the repository root folder for details.
+ */
+
+import { WbsElementStatus } from '../types/project-types';
+import { TimelineStatus } from '../types/work-package-types';
+
+/**
+ * This function calculates the end date.
+ * @param start the start date
+ * @param weeks number of weeks
+ * @returns the start date after the weeks have passed
+ */
+const calculateEndDate = (start: Date, weeks: number) => {
+  const end = new Date(start);
+  end.setDate(start.getDate() + weeks * 7);
+  return end;
+};
+
+/**
+ * This symbol function calculates the duration
+ * @param wps an array of work packages
+ * @returns the duration of the project in weeks
+ */
+const calculateDuration = (wps: any) => {
+  if (wps.length === 0) return 0;
+  if (wps.length === 1) return wps[0].duration;
+
+  let firstStart = wps[0].startDate;
+  let lastEnd = calculateEndDate(firstStart, wps[0].duration);
+
+  for (const wp of wps) {
+    if (wp.startDate < firstStart) firstStart = wp.startDate;
+    const end = calculateEndDate(wp.startDate, wp.duration);
+    if (end > lastEnd) lastEnd = end;
+  }
+  const durationMilliseconds = lastEnd.getTime() - firstStart.getTime();
+  const durationWeeks = durationMilliseconds / (1000 * 60 * 60 * 24 * 7);
+  return Math.round(durationWeeks);
+};
```

Comment:
This whole function would be much simpler if calculated via array functions. Get an array of start dates and apply min. Get an array of End dates and Apply max. Diff three min start and max end.